src/request.ts

```diff
@@ -710,9 +710,14 @@ class DatastoreRequest {
     query = extend(true, new Query(), query);
 
     const makeRequest = (query: Query) => {
-      const reqOpts: RequestOptions = {
-        query: entity.queryToQueryProto(query),
-      };
+      const reqOpts = {} as RequestOptions;
+
+      try {
+        reqOpts.query = entity.queryToQueryProto(query);
+      } catch (e) {
+        setImmediate(onResultSet, e);
```

Comment:
Why the `setImmediate`? could I bother you for a comment explaining so the next soul doesn't trip?

Reply:
Personal preference really. In my experience I've found unit testing callbacks that can be called early (and synchronously) to sometimes give weird results. I'm not married to it if you would prefer to go without it.

Reply:
If I understand correctly, this is addressing edge-cases discussed in [this post about zalgo](https://blog.izs.me/2013/08/designing-apis-for-asynchrony)?

Definitely a bit icky, as the language evolves, I think I'll be moving towards code more like this:

```
try {
  reqOpts.query = await entityToQueryProto(query); 
} catch (e) {
   onResultSet(e);
  return;
}
```

If a method could have an async error, treat it as an async method in general .... _but_, I can understand not wanting to take on this refactor.

I'd just suggest adding a comment, speaking to zalgo, e.g., _this ensures that the API doesn't sometime throw a synchronous error._

Reply:
Done! Thanks for the link to that post, pretty nice read!